### ISSUE TITLE
Flash reverted upon reboot

### ISSUE THREAD

**Issue by deeenes** [xy-tech/X2100-BIOS-EC#3]
Hello,

First, thank you so much for the updates, patches, and the this nice git repo and webpage with excellent support and guidance.

I am trying to update to the latest BIOS and EC, currently I am having the version from 13 July 2020. It seemingly goes alright with `flashrom`, however after reboot I always find the old BIOS in place. I followed the instructions exactly (incl. longer than 1 min wo power & battery). The BIOS image I'd like to flash is the ME disabled mSATA and the normal EC (no key swap, no fast charging). These below are referred to as `bios.bin` and `ec.bin`.

Checking if iomem=relaxed is set:
```
root@ # cat /proc/cmdline 
BOOT_IMAGE=../vmlinuz-linux init=/usr/lib/systemd/systemd rw iomem=relaxed initrd=../initramfs-linux.img
```

Combining the images:
```
root@ # ./x2100_helper.sh -c -b bios.bin -e ec.bin -o bios_ec.bin
4096+0 records in
4096+0 records out
4194304 bytes (4.2 MB, 4.0 MiB) copied, 0.00763375 s, 549 MB/s
64+0 records in
64+0 records out
65536 bytes (66 kB, 64 KiB) copied, 0.000522557 s, 125 MB/s
12224+0 records in
12224+0 records out
12517376 bytes (13 MB, 12 MiB) copied, 0.018895 s, 662 MB/s
Success!
Output binary is bios_ec.bin
```

Dumping the old rom contents:
```
root@ # flashrom -p internal -r dump_before.bin
flashrom dda1933 on Linux 5.15.11-arch2-1 (x86_64)
flashrom is free software, get the source code at https://flashrom.org

Using clock_gettime for delay loops (clk_id: 1, resolution: 1ns).
Cannot open file stream for /dev/mtd0
Found chipset "Intel Comet Lake U Premium".
Enabling flash write... SPI Configuration is locked down.
Enabling hardware sequencing because some important opcode is locked.
OK.
Found Programmer flash chip "Opaque flash chip" (16384 kB, Programmer-specific) mapped at physical address 0x0000000000000000.
Reading flash... done.
root@ # sha1sum dump_before.bin 
85176a67bd5c8696d999f1830ec2ef8f163a6ffb  dump_before.bin
```

Writing the new image:
```
root@ # flashrom -p internal -w bios_ec.bin 
flashrom dda1933 on Linux 5.15.11-arch2-1 (x86_64)
flashrom is free software, get the source code at https://flashrom.org

Using clock_gettime for delay loops (clk_id: 1, resolution: 1ns).
Cannot open file stream for /dev/mtd0
Found chipset "Intel Comet Lake U Premium".
Enabling flash write... SPI Configuration is locked down.
Enabling hardware sequencing because some important opcode is locked.
OK.
Found Programmer flash chip "Opaque flash chip" (16384 kB, Programmer-specific) mapped at physical address 0x0000000000000000.
Reading old flash chip contents... done.
Erasing and writing flash chip... Erase/write done.
Verifying flash... VERIFIED.
```

Dumping the new rom contents:
```
root@ # flashrom -p internal -r dump_after-flash.bin
flashrom dda1933 on Linux 5.15.11-arch2-1 (x86_64)
flashrom is free software, get the source code at https://flashrom.org

Using clock_gettime for delay loops (clk_id: 1, resolution: 1ns).
Cannot open file stream for /dev/mtd0
Found chipset "Intel Comet Lake U Premium".
Enabling flash write... SPI Configuration is locked down.
Enabling hardware sequencing because some important opcode is locked.
OK.
Found Programmer flash chip "Opaque flash chip" (16384 kB, Programmer-specific) mapped at physical address 0x0000000000000000.
Reading flash... done.
```

Comparing checksums:
```
root@ # sha1sum dump_after-flash.bin 
19d82627a4c2cdf16a861e3550de121dc21bf532  dump_after-flash.bin
root@ # sha1sum bios_ec.bin 
19d82627a4c2cdf16a861e3550de121dc21bf532  bios_ec.bin
```

Shutting down computer and checking the rom contents after reboot:
```
root@ # systemctl halt
# computer shuts down and boots up here
root@ # flashrom -p internal -r dump_after-reboot.bin
flashrom dda1933 on Linux 5.15.11-arch2-1 (x86_64)
flashrom is free software, get the source code at https://flashrom.org

Using clock_gettime for delay loops (clk_id: 1, resolution: 1ns).
Cannot open file stream for /dev/mtd0
Found chipset "Intel Comet Lake U Premium".
Enabling flash write... SPI Configuration is locked down.
Enabling hardware sequencing because some important opcode is locked.
OK.
Found Programmer flash chip "Opaque flash chip" (16384 kB, Programmer-specific) mapped at physical address 0x0000000000000000.
Reading flash... done.
root@ # sha1sum dump_after-reboot.bin 
1b0a90cceebde4709ba75b27a115dfb1e1fc5d63  dump_after-reboot.bin
```

Of course the checksum could be different due to other reasons, but I've checked inside the BIOS setup, and indeed the old version is there.

Some further things I experience:
- Although the BIOS is not updated, the settings are reset
- The computer boots very slow at the first boot after flashing:
  - Until 45s, only black screen, no backlight
  - At 45s, it restarts itself again
  - 8s later the NB51 logo appears
  - 4s later it restarts itself again
  - 8s later again the NB51 logo appears
  - 10s later the OS starts to boot up
 
Any help is very much appreciated.

**Comment by deeenes** [comment]
(Note: last time I updated it from Windows, but it's not an option any more, I have to do it from Linux)

**Comment by deeenes** [comment]
I tried two different BIOS versions to see if the issue is related to specific versions. In both cases the BIOS remained the 13 July 2020 version.

First I flashed these:
https://github.com/xy-tech/X2100-BIOS-EC/blob/2/x2100/bios_hap_0_safe.bin
https://github.com/xy-tech/X2100-BIOS-EC/blob/2/x2100/ec_patched.bin

After reboot the version shown in the BIOS was 13/07/2020, but the NVMe drive became invisible from the BIOS and I could boot only from a USB disk. From the OS the NVMe was visible and I could flash another BIOS, which made the NVMe available for boot again. It's interesting that the BIOS remains the same but we see a change in behavior after flashing.

Then I flashed these:
https://github.com/xy-tech/X2100-BIOS-EC/blob/main/x2100/bios/normal/bios_15122021.bin
https://github.com/xy-tech/X2100-BIOS-EC/blob/main/x2100/stock/ec_unpatched.bin

The BIOS version remained 13/07/2020 after reboot, I could boot from the NVMe drive.

In conclusion, no matter which version I flash, the BIOS remains unchanged or gets restored on reboot.

**Comment by deeenes** [comment]
After discussing with Jacky, it turned out that the BIOS version is not supposed to change, the new unofficial BIOS versions show 13/07/2020 just like the latest official one.

**Comment by xy-tech** [comment]
Sorry about the late response. Yea the date don't change because the BIOS date isn't changed in AMIBIOS when I generate it. However, most of the important changes are in the EC firmware and not in the BIOS so it's not too important.
